uwblueprint/sistering
# Set up backend and frontend deployment with CI

## Ticket link

Closes #67



## Implementation description
* Edited existing GitHub Actions workflow and Firebase config files to deploy frontend on PR creation and update (preview channel), and merge (main dev channel)
* Generated Firebase service account and configured in GitHub for deployment via Firebase CLI (see [here](https://firebase.google.com/docs/hosting/github-integration) for more info)
* Created new Heroku project for backend dev deployment and wrote GitHub Actions workflow for automated deploys on PR merges, see `.github/workflows/heroku-deploy-dev.yml`
* Deleted `firebaseServiceAccount.json` and modified Firebase SDK initialization to read environment variables directly
* Un-ignored `nodemailer.config.ts` after replacing secrets with env variables



## Steps to test
1. Click the frontend preview link commented by the github-actions bot in this PR. **You will see an alert with an error message, this can be ignored for now by clicking "OK" (Google login is not a critical feature, will debug later).**

![image](https://user-images.githubusercontent.com/30205929/145698692-de10d9b8-c2c8-4d7a-b219-845fbfd3d8e2.png)

2. Login with @sherryhli's test account credentials (check #sistering-dev in Slack). It may take a few seconds to get a response as Heroku dynos sleep after 30 minutes of inactivity.

![image](https://user-images.githubusercontent.com/30205929/145698747-40ace9dc-6fbe-4ff0-8cb7-69e617a38889.png)

3. Test features such as "Create Entity", "Update Entity", "Display Entities" through the frontend

4. You can also run GraphQL queries directly on the backend (URL is https://sistering-dev.herokuapp.com/graphql), but note that GraphQL Playground is not available in production. Try using a client like Altair or Postman

5. Ensure that GitHub Actions have all succeeded on this PR. Backend deployment on PR creation and update is disabled, but here are the [logs](https://github.com/uwblueprint/sistering/runs/4495361994?check_suite_focus=true) from my last run

6. After this PR is merged, ensure that frontend and backend deployment workflows are triggered



## What should reviewers focus on?
* Heroku deployment workflow: `.github/workflows/heroku-deploy-dev.yml`


## Checklist
